dataflow/src/main/java/org/checkerframework/dataflow/expression/ClassName.java

```diff
@@ -17,10 +20,28 @@
   /**
    * Creates a new ClassName object for the given type.
    *
-   * @param type the type for this ClassName
+   * @param type the type for this ClassName: declared, primitive, an array of declared or
+   *     primitive, or void
    */
   public ClassName(TypeMirror type) {
     super(type);
+    TypeMirror baseType;
+    TypeKind baseKind = type.getKind();
+    if (baseKind != TypeKind.ARRAY) {
+      baseType = type;
+    } else {
+      baseType = TypesUtils.getInnermostComponentType((ArrayType) type);
+      baseKind = baseType.getKind();
+    }
+    // Primitives, arrays, and void are permitted so FieldAccess can represent class literals.
```

Comment:
` FieldAccess can represent class literals.` I thought that `ClassName` represents class literals?

Reply:
ClassName represents class names, such as `int` or `String`.
The JLS uses the term "class literal" to mean a class name or array, followed by `.class`, such as `int.class` or `String.class` or `int[].class`.
Maybe you were thinking of "class literal" differently, eg as a class name?  Or did you think that ClassName represents `int.class` and `String.class`?

Reply:
> Or did you think that ClassName represents int.class and String.class?

Yes, that's what I think.  You also changed the Javadoc of `ClassName` to say that it represents class literals.  Also, the `toString`  method appends `".class"' to `typeString`.

Reply:
Oh I see that Javadoc says `"as part of ... a class literal"`.

Reply:
Currently `ClassName` are used for class literals, the whole class literal not just the "class name" part.  See:
https://github.com/typetools/checker-framework/blob/d6de8b7a0f66f35e70fa285b5b3f1def0969cc41/framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java#L825-L835

Reply:
Also here:
https://github.com/typetools/checker-framework/blob/d6de8b7a0f66f35e70fa285b5b3f1def0969cc41/dataflow/src/main/java/org/checkerframework/dataflow/expression/JavaExpression.java#L276-L281

Reply:
In that case, there are several other bugs (and missing/misleading documentation) that led me astray.
I opened https://github.com/typetools/checker-framework/pull/4636 to address it.